### ISSUE TITLE
persistent undo: handle very long paths, avoid E828

### ISSUE THREAD

**Issue by chrisduerr** [neovim/neovim#7073]
So recently when going through my plugins and cleaning stuff up I found that I still have https://github.com/pixelastic/vim-undodir-tree installed. I thought neovim might have solved that problem and tested it out. But neovim still has this problem with deep directory structures (or just a really long filename).

I feel like this is something Neovim could add by default because I do not see a drawback to it right now. If there is a drawback which I am not aware of, please let me know.

**Comment by justinmk** [comment]
I've not seen E828. What system do you see this on?

quoting https://github.com/pixelastic/vim-undodir-tree :

> The main goal of this plugin is to avoid the E828 error one can get in vim when trying to save a file with a very long file path and undofile enabled.

> instead of saving undo files as `%original%path%of%the%file`, it will create the `original/path/of/the/` folder structure inside of undodir, and save the undo file in it.

One (minor, I guess) drawback of this change is that nvim needs to crawl the tree instead of a flat list of files.

A different approach would be to compress the path and store the result as a base64-encoded filename.

**Comment by ZyX-I** [comment]
It would be more logical not to compress the path (it does not remove the limit, it only sometimes makes it larger), but use a hash. This has many advantages:

1. AFAIK we do not have any compression libraries linked yet. But we do have hash implementation.
2. Hash size is constant.

Cons is obviously that one will need to parse undo file to get the file name because hashing is a one-way path. And we will need to actually add the filename there, and provide some way to get it out: for scripting purposes, somebody may want to do smart clean-up.

**Comment by ZyX-I** [comment]
It would be more logical not to compress the path (it does not remove the limit, it only sometimes makes it larger), but use a hash. This has many advantages:

1. AFAIK we do not have any compression libraries linked yet. But we do have hash implementation.
2. Hash size is constant.

Cons:

1. One will need to parse undo file to get the file name because hashing is a one-way path. And we will need to actually add the filename there, and provide some way to get it out: for scripting purposes, somebody may want to do smart clean-up.
2. To be safe it is needed to handle collisions, file path in the undo file will also come handy. Theoretically this will also destroy 2. from pros, but in practice I doubt you will ever get even 100 sha256 collisions on file names on a single machine.

**Comment by chrisduerr** [comment]
This happens on linux if you have a huge amount of folders with long names inside each other.
I just thought that this seems exactly like the kind of thing NeoVim would change over Vim. However I am not sure about the advantages and disadvantages. I just know that this issue exists and the plugin fixes it for me.

**Comment by justinmk** [comment]
> I just thought that this seems exactly like the kind of thing NeoVim would change over Vim

Nothing's blocking it, but it still needs a human to think about it carefully, write tests, and create a patch. Not to mention I've never heard of this issue until now.

**Comment by chrisduerr** [comment]
Yeah I understand that. I just haven't seen it mentioned anywhere so I thought I'd at least try to start the conversation. I would be down to look into it but I don't have much experience so it might take some time until I can play around with it.

**Comment by KillTheMule** [comment]
> I would be down to look into it but I don't have much experience

There's a gitter channel where people generally are helpfull and interested. Join us and we'll help you start :)

**Comment by appelgriebsch** [comment]
Any updates on this? Happens to me multiple times in my work projects...

**Comment by justinmk** [comment]
Same answer as before: https://github.com/neovim/neovim/issues/7073#issuecomment-328358428

**Comment by fingust** [comment]
Just ran into the same issue when opening a large project. I tried installing `pixelastic/vim-undodir-tree` but it didn't solve the issue. I guess not compatible with neovim? I am running on Linux and ext4. I think the issue is the fact that I have home folder encryption (I have to, because ISO 27001) which for some reason vastly reduces max filename length (from 255 to 143). But I can see large projects having issues anyway. The only "solution" seems to be to disable the `undofile` or just deal with it :/

**Comment by awilkins** [comment]
Running into this one editing markdown files with comments in then, in a big directory tree synced from Sharepoint - Windows ecosystem users seem to love their needlessly wordy and space-separated directory paths, even if OneDrive will choke on them when you hit 240 characters deep.

> Windows, where the entire path has a very small limit.

The really stupid part is that it's not a limit of NTFS, which allows path lengths up to 64kB long ... it's a holdover from FAT, but they kept the assertions against long paths in most of the Windows API.

I've had the Team Foundation Server refuse to serve me files because it thinks I would be putting them in a path that would be nested too deep (even if you patch the limit out of the client ...). OneDrive just refuses to sync whole folders until you remove the one file that takes the path over the limit.

Anyway, we're not going to fix Windows here. But I'd propose that rather than being a configurable option, why not have the path-hashing behaviour as the fallback behaviour in the event that a tempfile name would trigger E828? This retains the previous naming behaviour (which is presumably regarded as a feature) for as many cases as possible, but preserves being able to have undo for paths that exceed the limit.

**Comment by justinmk** [comment]
This is unblocked now that we have `vim.fs.slug()` https://github.com/neovim/neovim/pull/41005 ; just need to migrate undofile to use it.
